src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,646 @@
+					  "stopping already started services and pg_autoctl",
+					  service->name);
+
+			for (idx = serviceIndex - 1; idx > 0; idx--)
+			{
+				if (kill(services[idx].pid, SIGQUIT) != 0)
+				{
+					log_error("Failed to send SIGQUIT to service %s with pid %d",
+							  services[idx].name, services[idx].pid);
+				}
+			}
+
+			/* we return false always, even if supervisor_stop is successful */
+			(void) supervisor_stop(pidfile);
+
+			return false;
+		}
+	}
+
+	/* now supervise sub-processes and implement retry strategy */
+	if (!supervisor_loop(start_pid, services, serviceCount, pidfile))
+	{
+		log_fatal("Something went wrong in sub-process supervision, "
+				  "stopping now. See above for details.");
+		success = false;
+	}
+
+	return supervisor_stop(pidfile) && success;
+}
+
+
+/*
+ * service_supervisor calls waitpid() in a loop until the sub processes that
+ * implement our main activities have stopped, and then it cleans-up the PID
+ * file.
+ */
+static bool
+supervisor_loop(pid_t start_pid,
+				Service services[],
+				int serviceCount,
+				const char *pidfile)
+{
+	int subprocessCount = serviceCount;
+	int stoppingLoopCounter = 0;
+	bool shutdownSequenceInProgress = false;
+	bool cleanExit = false;
+
+	/* wait until all subprocesses are done */
+	while (subprocessCount > 0)
+	{
+		pid_t pid;
+		int status;
+
+		/* Check that we still own our PID file, or quit now */
+		(void) check_pidfile(pidfile, start_pid);
+
+		/* If necessary, now is a good time to reload services */
+		if (asked_to_reload)
+		{
+			int serviceIndex = 0;
+
+			for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+			{
+				Service *service = &(services[serviceIndex]);
+
+				log_debug("Reloading pg_autoctl %s service", service->name);
+
+				(void) (*service->reloadFunction)(service);
+			}
+
+			/* reset our signal handling facility */
+			asked_to_reload = 0;
+		}
+
+		/* ignore errors */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				if (errno == ECHILD)
+				{
+					/* no more childrens */
+					if (asked_to_stop || asked_to_stop_fast)
+					{
+						/* off we go */
+						log_info("Internal subprocesses are done, stopping");
+						return true;
+					}
+				}
+				else
+				{
+					log_fatal("Failed to call waitpid(): %m");
+					return false;
+				}
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 */
+				if (asked_to_stop || asked_to_stop_fast)
+				{
+					cleanExit = true;
+
+					/*
+					 * Stop all the services.
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(services,
+															subprocessCount);
+					}
+
+					++stoppingLoopCounter;
+
+					if (stoppingLoopCounter == 1)
+					{
+						log_info("Waiting for subprocesses to terminate.");
```

Comment:
I think the intention was that with some luck we proceed on the first loop to find exited processes and then we will never have to display the "waiting" message to the user. It would be too fast for that...

Reply:
If that's the intention then it doesn't do that, since the `stoppingLoopCounter` is incremented right before this check. So it's effectively the same check as `stoppingLoopCounter == 0` before the increment (that's what I tried to indicate with my previous comment, but that was not very clear).

I'm not sure if the intention you mention is indeed what we want, since we will start waiting as soon as we call `supervisor_stop_process`. But if you disagree I'm fine with that.

Reply:
Ah yeah I see what you mean, and feel a little stupid about that. With the refactoring it should now be okay, unless I had another brain fart.